### PR TITLE
v2.4.0 update

### PR DESCRIPTION
**Pull request by blt950**
**Changelog**
**Remember to migrate database**

- Made training statistics (for respective area) available for moderators as well
- Added functionality for moderators to edit training requests (e.g. changing applied ratings)
- Added training type and english only options to manual training creation (Thanks Christian)
- Added pass/fail rate for CTPs graph to training report
- Added "workmail" functionality for moderators. You can now in your settings select a VACC mail to send all CC emails to instead of your private one. It automatically expires after 30 days, and 7 days prior you'll be asked in dashboard to extend.
- Added access report for moderators, this gives a list of everyone with access to Control Center. Easier to manage and see who to remove access from.
- Added audit logs for admins. This includes more detail logs of actions done in CC. Together with this update we've updated the quality and log details, though older logs before this day will display in old format and not categorised.

- Training reports and exams are now sorting chronologically, exams are no longer only on top
- Fixed global search window opening when searching in tables
- Fixed date sorting order in vatbooks dropdown selector
- Fixed a bug where "english only" flag did not save from applications, this might explain all new requests having wrong flag
- Fixed missing permission check when creating manual training requests

- Updated charts.js to new major v3 version
- Deploy.sh chmod (Thanks Christian)

### CODE REVIEW

File: resources/views/vote/show.blade.php [hunk 1]
```diff
@@ -110,7 +110,7 @@
                 intersect: false
             },
             scales: {
-                yAxes: [{
+                y: [{
```
Comment (gustavkauman) [inline resources/views/vote/show.blade.php:113]: I don't know the JS library, but I was just wondering... In `reports.trainings` the scales are made with `x` and `y` with pure JS objects, not wrapped in an array. Is it intentional that it's still wrapped in an array here?

File: app/Http/Controllers/TrainingController.php [hunk 2]
```diff
@@ -262,8 +264,22 @@ public function show(Training $training)
...
+        $reports = TrainingReport::where('training_id', $training->id)->get();
+
+        $reportsAndExams = collect($reports)->merge($examinations);
+        $reportsAndExams = $reportsAndExams->sort(function ($a, $b) {
+
+            // Define the correct date to sort by model type is report or exam
+            is_a($a, '\App\Models\TrainingReport') ? $aSort = Carbon::parse($a->report_date) : $aSort = Carbon::parse($a->examination_date);
+            is_a($b, '\App\Models\TrainingReport') ? $bSort = Carbon::parse($b->report_date) : $bSort = Carbon::parse($b->examination_date);
+
+            // Sorting algorythm
```
Comment (gustavkauman) [inline app/Http/Controllers/TrainingController.php:277]: Spelling mistake :stuck_out_tongue:

File: app/Policies/TrainingPolicy.php [hunk 3]
```diff
@@ -109,6 +109,17 @@ public function create(User $user)
...
 
+    /**
+     * Check if the user has access to edit a training details
+     *
+     * @param User $user
+     * @return bool
+     */
+    public function edit(User $user)
+    {
+        return $user->isModeratorOrAbove();
```
Comment (gustavkauman) [inline app/Policies/TrainingPolicy.php:120]: This should take the area as well, so only a moderator in the given country update the training request.

File: app/Policies/ManagementReportPolicy.php [hunk 4]
```diff
@@ -30,13 +30,13 @@ public function viewMentors(User $user) {
         return $user->isModeratorOrAbove();
     }
 
-    /**
+        /**
```
Comment (gustavkauman) [inline app/Policies/ManagementReportPolicy.php:33]: Can this one line be reverted? :stuck_out_tongue:

File: resources/views/reports/access.blade.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+                                <th data-field="name" data-sortable="true" data-filter-control="input">Name</th>
+                                @foreach($areas as $area)
+                                    <th data-field="access-{{ $area->id }}" data-sortable="true" data-filter-control="input">Access {{ $area->name }}</th>
+                                @endforeach
+                            </tr>
+                        </thead>
+                        <tbody>
+                            @foreach($users as $user)
+                            <tr>
+                                <td><a href="/user/{{ $user->id }}">{{ $user->id }}</a></td>
```
Comment (gustavkauman) [inline resources/views/reports/access.blade.php:33]: This should be done with the `route()` helper instead, so we're sure it doesn't break should we change the route

**Review by gustavkauman — CHANGES_REQUESTED** [review]
Check the inline comments. Great new features!

File: app/Console/Commands/UpdateWorkmails.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+     * @return int
+     */
+    public function handle()
+    {
+        
+        // Check for expired workmails
+        DB::table('users')->where('setting_workmail_expire', '<=', date('Y-m-d H:i:s'))->update(['setting_workmail_address' => NULL, 'setting_workmail_expire' => NULL]);
+
+        // Check for users that no longer hold a moderator or admin rank
+        foreach(User::whereNotNull('setting_workmail_address')->get() as $user){
```
[the remark this answers is not in the document]
Comment (blt950) [reply app/Console/Commands/UpdateWorkmails.php:47]: Ref. Discord discussion. Only expire when removing rank is not a good enough insurance the workmail is correct.

File: resources/views/vote/show.blade.php (hunk 1 above)
Comment (blt950) [reply resources/views/vote/show.blade.php:113]: Good catch, solved in push. Also forgot to init charts on this view after the change to charts.js v3 in previous commits.

**Comment by blt950** [comment]
Issue should be ironed out, and I tested it a bit and got rid of some other bugs :)

**Pull request merged**